SimpleUI Application

kb1lqc opened the pull request:
I created a simple user interface. No frills. It's pure Python Flask + Javascript/HTML. There's a decent amount of new code but it all boils down to the UI and whether it works or not.

- It assumes basic interfacing for simplicity in development... i.e. uses a configuration file to set callsigns instead of user input via web-browser as that adds a bit of code that overall isn't the mission of this application.
- ADCs are scaled to volts at the ADC pin in javascript for the UI but saved in the database as bits for posterity. I did not indicate units on the UI. Let me know if they really should be added.

@reillyeon I'd appreciate a quick review. I don't think we need a super in-depth review here for Javascript/HTML but I did set up JSHint and htmllint locally with node.js to check to what I believe the JQuery project requires as we aim to mimic in CONTRIBUTIONS.MD. We do not currently have any automated node.js testing so I just left the configuration documents in for local installs.

@kb1lqd I don't think you need to dive into the review if you don't want to so I'll base my merge off of Reilly's comments unless say otherwise.

The readme.md shows the basic configuration (minimal) and even a gif of the UI working. This is sufficient to perform a basic Amateur Radio High Altitude Balloon Mission as well as general telemetry/command interfacing needs. There are some known issues still present as described in #145 that I believe should not hold up this PR as they could be firmware issues unless proven not to be.

![SimpleUI](https://github.com/kb1lqc/Faraday-Software/raw/simple_web_ui/Applications/SimpleUI/images/simpleuibasic.gif)

kb1lqc committed:
Initial Simple UI Application

Framework code for #145 to help Rose-Hulman Institute of Technology have a
simple user interface. This is a barebones flask server which responds
with an HTML template.

kb1lqc committed:
Merge branch 'master' into simple_web_ui

kb1lqc committed:
Load JQuery library from index.html

Added javascript call in HTML document to load the JQuery library. This
also included a few folder additions and html file formatting changes.

kb1lqc committed:
Moved to layout and index html files

Using layout.html to start up javascript and basic HTML format. The
index.html file serves up jinja code.

kb1lqc committed:
Moved python to server index.html, added alert

Now python properly serves an alert every 5 seconds as well as a console
log test in index.html. This is preparing for ajax calls to obtain
telemetry.

kb1lqc committed:
Removed Console Log Test

Removed test code to log a "test" to console.

kb1lqc committed:
Added getJSON() JQuery Function

Added $.getJSON() to the index.html and temporarily search a placeholder
website that returns JSON. I also commented out the setInterval code.

kb1lqc committed:
Moved getJSON to function and used interval

I moved the getJSON code to it's own function getTelemetry(). Then I moved
the setInterval code and have it calling getTelemetry once per second to
update with new data.

kb1lqc committed:
Working CORS Requests to get JSON Telemetry

If CORS is enabled on the telemetry server then data will be obtained for
the latest packet. No regard to station has been made, the latest packet
is returned and printed to console.

kb1lqc committed:
Basic JQuery updated user interface

Implemented CORS based automated JQuery updating user interface
HTML/Javascript. Twice per second the Telemetry server is queried for data
and updated to show the latest station data. Only a subset of information
is currently shown and this should be able to specify which station to
listen to.

kb1lqc committed:
Added Bootstrap Javascript library

Added bootstrap Javascript library and pulled it into the simpleUI
application. This will allow basic layout formatting and styling.

kb1lqc committed:
Correct Javascript/CSS File Loading

I incorrectly loaded in CSS files as javascript which didn't work. Only
when I used an HREF link (duh`) did it actually work and my "test" and
"ing" correctly move into bootstrap grids.

kb1lqc committed:
Added Bootstrap Panels

Implemented bootstrap CSS panels as well as some other text to identify
data groupings. Placed station an epoch into a css well also.

kb1lqc committed:
Fixed up Bootstrap Formatting

Centered and aligned text as well as added heading formatting.

kb1lqc committed:
Converted Epoch to Date

Used javscript to convert epoch into a local time date object. Removed
epoch text and just printing the raw date object to HTML.

kb1lqc committed:
Added ability to specify station

Used URL parameters to pass data to the javascript specifying a station to
listen for.

kb1lqc committed:
Formatted Callsign

Top left corner where callsign is placed was updated to be bold.

kb1lqc committed:
Converteed Knots to m/s

Converted speed from GPS in knots into speed in meters per second. Used
javascript.

kb1lqc committed:
Check for both callsign and nodeid

If both callsign and nodeid are present in url parameters the simpleui is
returned. Otherwise an error is returned indicating the correct formatting
of the URL.

kb1lqc committed:
Added warning text when station note found

Added a simple text indication when a station cannot be found in telemetry
database. The necessary logic provided to detect a fault/erroneous SQLite
lookup and respond with the appropriat text.

kb1lqc committed:
Added hyphen to error callsign nodeid

There was a missing hyphen between the callsign and nodeid when a station wasn't
found and text indicated this. Added.

kb1lqc committed:
Removed flake8 two line error

Removed error when expecting 2 lines but got one after a function. Did not
fix CORS error in telemetry as that's handled in another PR.

kb1lqc committed:
Added IO Signals, commented out HAB

Added GPIOSTATE, RFSTATE, and IOSTATE signals to user interface and
commented out the HAB items. I will add the HAB stuff back in shortly on a
new row. Currently no way to test all IO as some other firmware/software
errors prevent me from seeing all telemetry or test.

kb1lqc committed:
Added HAB Timer Back In

Added the High Altitude Balloon timer back into the UI and correctly
formatted it on the screen. There is still a javascript error that is
present...

kb1lqc committed:
Commented out IOState Items

Since Issue #151 exists this will fail causing an error. Removing for now
but leaving code in for easy implementation when fixed.

kb1lqc committed:
Added simple GPIO command radio buttons

Implemented simple radio buttons for GPIOs. I also had to send down
callsigna and nodeid info with the form POST data to correctly redirect.
This implements the POST function in simpleui flask server now.

kb1lqc committed:
Some basic ability to command LEDs

Only manual on/off but something. Stopping for the night. Need to figure
out a path forward soon.

kb1lqc committed:
Initial Local and Remote Commanding Added

Local commanding seems to work but remote commanding is not. Need to
investigate why not.

kb1lqc committed:
Merge branch 'master' into simple_web_ui

Pulled in telemetry CORS update from master into simple_web_ui branch.

kb1lqc committed:
Most Remote commands now work

Added remote command if statements to detect UI form input for remote
station commands. I say this mostly works because the command
library/allocations is actually erroneously mapped so it's hard to test.
MOSFET HAB cutdown event remote commanding doesn't seem to work either.

kb1lqc committed:
Added local HAB Timer Commanding

Added a few local commands for High Altitude Balloon timer use.

kb1lqc committed:
Added HAB Timer to UI Page

Added the HAB timer count to the user interface.

kb1lqc committed:
Added Remote HAB Timer Control

Still appears not to work but the base code is there to fix it when we
root cause why the MOSFET/HAB Timer doesn't seem to work over RF.

kb1lqc committed:
Rounded GPS speed to two decimal places

Added javascript rounding of GPS speed to two decimal places.

kb1lqc committed:
Added space between boardtemp and ADC data in UI

kb1lqc committed:
Removed unecessary HTML form table

Cleaned up local and remote command dropdown menues by removing wierd HTML
table.

kb1lqc committed:
Added boilerplate RF station input

Form will eventually input user defined callsign. For now it's
boilerplate.

kb1lqc committed:
Removed need for URL callsign for commanding

I decided that it's obnoxious at this time to perform URL based local and
remote station identification. It's much easier to specify the stations in
a config file and move on.

kb1lqc committed:
Removed need for URL station identification

A GET query now simply reads the configuration file to see the local and
remote stations.

kb1lqc committed:
Separated Telemetry displayed from station command

I now specify local/remote stations for commanding but additionally
specify the station to receive telemetry from.

kb1lqc committed:
Fixed all currently reported Flake8 errors

Changed all currently checked Flake8 errors. Also, I removed some
unecessary code as I saw.

kb1lqc committed:
Cleaned up and commented code a bit

kb1lqc committed:
Updated Javascript comments

Commented on Javascript to help identify sections of code for easier
reading.

kb1lqc committed:
Changed redirect to just go to root page

kb1lqc committed:
Updated readme to include basic config and use

kb1lqc committed:
Merge branch 'master' into simple_web_ui

kb1lqc committed:
Moved simpleui.py to new faraday proxyio location

Faraday proxyio files have moved as we move towards modularizing the
project. This also means that my proxyio.py file had to insert a new path
and use faraday.proxyio as the "module" to import items such as
faradaybasicproxyio.

kb1lqc committed:
Parse RFSTATE into 8-bit binary number & update

Correctly parse RFSTATE into an array of binary numbers even if all zeros.
This is then updated on specific parts of the UI with JQuery.

kb1lqc committed:
Updated IOSTATE parsing and display

Much like RFSTATE this commit properly parses IOSTATE into an array of
binary items that is displayed appropriately.

kb1lqc committed:
Updated GPIO parsing and display

Used the same method as RFSTATE and IOSTATE to parse GPIOSTATE correctly.

kb1lqc committed:
Moved GPIOSTATE, RFSTATE, and IOSTATE parsing

Parsing is now performed prior to any IO state output display in
preparation for a reorganization of the display code.

kb1lqc committed:
Organized IO display

Reorganized how the IO section is computed and displayed.

kb1lqc committed:
Fixed all known HAB and MOSFET remote Non-working

With these changes to commands called the HAB timer and MOSFET appear to
be working on the remote station. Verification was performed with HAB and
MOSFET states and not with hardware testing (multimeter) yet. This should
fix #153.

kb1lqc committed:
Removed print statements and fixed formatting

Removed unecessary debug print statements and fixed Flake8 formatting
errors.

kb1lqc committed:
Added logging debug info to commands

When sending a local or remote command the corresponding action is logged
as a debug statement now. This could help in future tasks such as
debugging failed commands or logging when commands were sent for post
analysis.

kb1lqc committed:
Updated debug info and command dropdowns

Debug info now contains the specified stations which they are commanding.
Also, the command dropdown menue now specifies the local or remote text in
the button.

kb1lqc committed:
Removed unused hidden HTML form items

Old callsign test form items were removed. I've since moved to simply
reading in a configuration file for time savings on development work.
Tested and I can still command local and remote units.

kb1lqc committed:
Moved Javascript into it's own file

Javascript moved into it's own simpleui.js file and linked to in the
index.html. I also removed an unused piece of Javascript from index.html.

kb1lqc committed:
Adding grunt files

grunfile and package JSON configuration have been added for javascript
linting and other automated processes.

kb1lqc committed:
Extra bootstrap theme deleted

Example bootstrap theme HTML file deleted.

kb1lqc committed:
Updated Gruntfile to perform jshint and html error

HTML title added to both index.html and layout.html to mitigate an error
in htmllint. I should actually ignore it in index.html but moving on...
Added support for jshint.

kb1lqc committed:
Simpleui.js now passes jslint

kb1lqc committed:
Updated grunt linting for javascript

Used the suggested linting configuration for JQuery on the JSHint
configuration. All observed errors were fixed and known warnings that were
actually OK were marked as ignore.

kb1lqc committed:
Removed reverse of GPIO field to fix

GPIOs 3 through 8 now work as expected via telemetry.

kb1lqc committed:
Removed GPIO9 as it doesn't exists

Woops

kb1lqc committed:
Added documentation image and text for use

Added a gif of the SimpleUI being used as well as some descriptive text on
the readme.md.

kb1lqc committed:
Deleted unused reverse

GPIO reverse was wrong, though, otherse for RFSTATE and IOSTATE appear to
be correct. I deleted the commended out line.

kb1lqc committed:
Fixed gif image error

I missed the "!" with markdown for displaying images. Added to correctly
show the screenshot of the UI being used.

kb1lqd commented on Applications/SimpleUI/simpleui.ini:
```diff
@@ -0,0 +1,19 @@
+[FLASK]
```
You committed your personal simpleui.ini when is should be added to GitIgnore.

kb1lqc replied:
Removed

kb1lqd reviewed (changes requested):
Reviewing gpio allocations shows the need to change PORT 3 GPIO allocations since they are backwards from firmware:

https://github.com/kb1lqc/Faraday-Software/blob/simple_web_ui/faraday/proxyio/gpioallocations.py#L11

**Firmware Allocations**
https://github.com/FaradayRF/Faraday-Firmware/blob/swissue153/REVA_Faraday.h#L65
![image](https://cloud.githubusercontent.com/assets/339144/25200766/55d2195c-2504-11e7-8e5f-d8ab3fe64e26.png)


**Software Allocations**
https://github.com/kb1lqc/Faraday-Software/blob/simple_web_ui/faraday/proxyio/gpioallocations.py#L11

![image](https://cloud.githubusercontent.com/assets/339144/25200835/9fd776dc-2504-11e7-9f5d-6784dd6c1bfe.png)

kb1lqd commented on Applications/SimpleUI/static/simpleui.js:
```diff
@@ -0,0 +1,103 @@
+                    rfstate.reverse();
+
+                    // Populate GPIO data in IO section
+                    $(".GPIO0").text(gpio[0]);
```
(With updated firmware per my firmware issue ticket)

When I command GPIO 7 I see GPIO 0 turn on:
![image](https://cloud.githubusercontent.com/assets/339144/25201025/55a8040e-2505-11e7-86ba-deaabbfe9810.png)

Reviewing this in code composer shows that I properly bit shifted GPIO 7 into the MSB location that matches the excel document:

![image](https://cloud.githubusercontent.com/assets/339144/25201070/7e82305c-2505-11e7-900e-b65062c75780.png)

![image](https://cloud.githubusercontent.com/assets/339144/25201098/961cfaee-2505-11e7-8ad0-6d613a530a67.png)

@kb1lqc please review you're code here to make sure that you are properly allocating the bitfield.

kb1lqd replied:
Checking GPIO 4 shows I properly bitshift but you show GPIO 3 as changed:

![image](https://cloud.githubusercontent.com/assets/339144/25201196/cc3fddc6-2505-11e7-83ff-24cbc6e62426.png)

![image](https://cloud.githubusercontent.com/assets/339144/25201208/d0c0ae66-2505-11e7-8b28-b4a05752872f.png)

kb1lqd replied:
This actually makes sense because I send the telemetry down in little endian format but my parsing forced big endian:
https://github.com/kb1lqc/Faraday-Software/blob/simple_web_ui/faraday/proxyio/telemetryparser.py#L18

Swapping `<` for `>` didn't seem to fix it but it's a good clue.
could javascript also be forcing an endianess?

@Bryce please verify and see if you can fix, this is something trivial. If it's easier for me to just swap my bitfield around on Faraday let me know.

kb1lqc replied:
Using the [new firmware](https://github.com/FaradayRF/Faraday-Firmware/issues/72#issuecomment-295424504) I can clearly see the following when I print out the GPIO state in javascript from SimpleUI:

GPIO7 HIGH
![image](https://cloud.githubusercontent.com/assets/331540/25217662/99b6d79a-255c-11e7-8e5f-08f8127df73a.png)

GPIO0 HIGH
![image](https://cloud.githubusercontent.com/assets/331540/25217749/ea62e85a-255c-11e7-9014-0e6c72630d34.png)

GPIO1 HIGH
![image](https://cloud.githubusercontent.com/assets/331540/25217944/b7ce74bc-255d-11e7-8135-607e71e367dd.png)


Given the documentation showing the following GPIO State packet I believe this is simple reversed in the Javascript array:
![image](https://cloud.githubusercontent.com/assets/331540/25217821/2dcbf8e8-255d-11e7-9707-d6eac265b1f2.png)

However... Measuring with a multimeter to confirm GPIO output state...

# Commanding GPIOs
It appears commanding is messed up. Here's what happens when I send commands to Faraday GPIO

GPIO0: D2
GPIO1: D1
GPIO2: D0
GPIO3: D3
GPIO4: D4
GPIO5: D5
GPIO6: D6
GPIO7: D7
GPIO8: D8

_Clearly Commanding GPIO0 through GPIO2 is incorrect_

# GPIO Telemetry
D0: GPIO7
D1: GPIO6
D2: GPIO5
D3: GPIO4
D4: GPIO3
D5: GPIO2
D6: GPIO1
D7: GPIO7
D8: GPIO8

_Clearly GPIO0 never turns on and D0-D6 telemetry is reversed on SimpleUI_

kb1lqc replied:
After reversing GPIO in javascript with `gpio.reverse();` I obtain the following:

D0: GPIO0
D1: GPIO1
D2: GPIO2
D3: GPIO3
D4: GPIO4
D5: GPIO5
D6: GPIO6
D7: GPIO7
D8: GPIO8

AWESOME @kb1lqd ! Telemetry now corresponds to MSP430 pins (D0-D8 are silkscreen designators on the PCBA). So we've fixed 

- The telemetry coming down UART is now correctly indicating GPIO values
- SimpleUI correctly outputs GPIO values to the user

Not sure if we really need to care about endianness at the moment. My Javascript code to create the array performs an into the binary string + shift already which actually might be the source of the flip. Regardless it's working right now and we have one issue in the way:

-[ ] GPIO0-GPIO2 commands are flipped

kb1lqc replied:
@kb1lqd so it appears as though from a commanding standpoint either the firmware is flipped or [GPIO allocations](https://github.com/kb1lqc/Faraday-Software/blob/simple_web_ui/faraday/proxyio/gpioallocations.py#L11) are flipped. Thoughts?

It appears as though GPIO allocations are ordered correctly so I would suspect commanding in firmware?

kb1lqd reviewed (changes requested):
Added review comments on GPIO bitfield endianess.

kb1lqc committed:
Added SimpleUI config file to gitignore

SimpleUI configuration file simpleui.ini added to gitingore to force the
use of changing simpleui.sample.ini over for each user.

kb1lqc committed:
Removed simpleui.ini from Git

Not needed and is user specific. Deleting per @kb1lqd suggestion on PR #166.

kb1lqc committed:
Reversed GPIO array to correct output

GPIO values now correctly display hardware output values. However,
commanding of GPIO0-GPIO2 is still reversed.

kb1lqc committed:
Removed second GPIO8 dropdown item

Duplicate removed from local list.

kb1lqc committed:
Updated .gitingore to ignore nodejs folder

node_modules was added when nodejs modules were installed. These are
however huge and should just be installed locally as they are debug items.

reillyeon committed:
Add grunt-contrib-jshint to package.json

The Gruntfile.js references this package but it was not listed in
package.json.

reillyeon committed:
Run Grunt tasks in .travis.yml

kb1lqc committed:
Merge pull request #1 from reillyeon/simple_web_ui

Enable HTML and JavaScript linting in Travis CI

reillyeon reviewed (commented):
This change looks good to me though I defer to @kb1lqd for final approval because he is more familiar with the GPIO assignments which appear to be a sticking point.

There are a couple of things I would like to see eventually happen in this code:

* Build more of the UI and form handling logic with code so that it is less repetitive. Copy-pasted code tends to lead to errors.
* Use Grunt/npm to pull in our dependencies like Bootstrap and jQuery during the build process so that we don't have to check them into git.

kb1lqc commented:
@reillyeon agreed and there was no real long-term thought put into this. RHIT is launching a balloon in a week or two and needed a user interface. This is no-frills. Ideally I'd like to use OpenMCT as it really seems to hand "missions" and situational awareness well. However that's really aimed at telemetry/command and being internet based I'd like to see another dashboard or similar project be utilized.

kb1lqd commented:
Firmware GPIO allocations match hardware:
![image](https://cloud.githubusercontent.com/assets/339144/25323408/e3f8c2fe-2873-11e7-8ea8-33aef26b1bc2.png)


The Python GPIO allocations do NOT match hardware in GPIO 0-2 which is shown in the swapped behavior during commanding. Please swap GPIO 0 with BIT 2 and GPIO 2 with BIT 0 as shown in the image below.

Update: https://github.com/kb1lqc/Faraday-Software/blob/simple_web_ui/faraday/proxyio/gpioallocations.py#L11

as shown: 
![image](https://cloud.githubusercontent.com/assets/339144/25323515/90920c8c-2874-11e7-807e-c88a4af1fbca.png)

Commit:
Swapped GPIO0-GPIO2 in FaradayIO gpioallocations.py which resolved the 0-2 Digital Out commanding flip

kb1lqd commented:
I pushed the GPIO swap to fix the last GPIO commanding issue. All looks good. @kb1lqc
